### PR TITLE
Bigtests

### PR DESCRIPTION
**Pull request by danmelamed**
alpha release of big_tests module

### CODE REVIEW

**Comment by JohnLangford** [comment]
When I type 'make', I get:
11:24AM humpty-10: make       ~/programs/vowpal_wabbit/big_tests [jl/ttypts/24]
bash: -c: line 1: syntax error: unexpected end of file
make: *** [valid] Error 1

Do you know what's up?

**Comment by danmelamed** [comment]
Nope ☺ .  When I try the same command, make does nothing because the file big_tests already exists.  What directory are you running from, and with which version of make and bash? If you want to run the default makefile in a directory other than the cwd, I believe the correct syntax is
make –C ~/programs/vowpal_wabbit/big_tests
Can you try `make` while cwd is big_tests?  Or if you tell me the full name of humpty-10, I can try to log in and look around.

Dan

**Comment by JohnLangford** [comment]
This is my home desktop which is behind a firewall---difficult to login to.

I'm running with cwd = ~/programs/vowpal_wabbit/big_tests  All commands 
are below.  I suspect some weirdo line ending problem? What environment 
are you building on?  This is:
12:00PM humpty-38: cat /etc/issue                              ~ 
[jl/ttypts/24]
Ubuntu 14.04.3 LTS \n \l

-John

11:58AM humpty-25: pwd        ~/programs/vowpal_wabbit/big_tests 
[jl/ttypts/24]
/home/jl/programs/vowpal_wabbit/big_tests
11:58AM humpty-26: make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for x86_64-pc-linux-gnu
11:58AM humpty-27: bash --version
GNU bash, version 4.3.11(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later 
<http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
11:58AM humpty-28: make       ~/programs/vowpal_wabbit/big_tests 
[jl/ttypts/24]
bash: -c: line 1: syntax error: unexpected end of file
make: *** [valid] Error 1

**Comment by danmelamed** [comment]
Turns out .ONESHELL: doesn’t work with make < 3.82, and I’m also bumping into other issues with make 3.81.  Since a lot of people are still working with 3.81, I’ll have to rework a few things….

**Comment by JohnLangford** [comment]
Ok.

-John

**Comment by danmelamed** [comment]
OK, please try again.

**Comment by JohnLangford** [comment]
This seems to run but fail on the first test now.  Something seems off because the feature counts differ?

 7:04AM humpty-67: make       ~/programs/vowpal_wabbit/big_tests [jl/ttypts/24]
Starting to run test 1a
--2015-12-22 07:04:47--  http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz
Resolving yann.lecun.com (yann.lecun.com)... 128.122.47.89
Connecting to yann.lecun.com (yann.lecun.com)|128.122.47.89|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 28881 (28K) [application/x-gzip]
Saving to: ‘train-labels-idx1-ubyte.gz’

100%[======================================>] 28,881      --.-K/s   in 0.008s  

2015-12-22 07:04:47 (3.40 MB/s) - ‘train-labels-idx1-ubyte.gz’ saved [28881/28881]

--2015-12-22 07:04:47--  http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz
Resolving yann.lecun.com (yann.lecun.com)... 128.122.47.89
Connecting to yann.lecun.com (yann.lecun.com)|128.122.47.89|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9912422 (9.5M) [application/x-gzip]
Saving to: ‘train-images-idx3-ubyte.gz’

100%[======================================>] 9,912,422   6.88MB/s   in 1.4s   

2015-12-22 07:04:48 (6.88 MB/s) - ‘train-images-idx3-ubyte.gz’ saved [9912422/9912422]

/home/jl/programs/vowpal_wabbit/big_tests/../vowpalwabbit/vw --oaa 10 -d /home/jl/programs/vowpal_wabbit/big_tests/dataSets/mnist.dir/train.prep -f mnist.model -b 24 --adaptive --invariant --holdout_off -l 0.1 --nn 40 --passes 24 -k --compressed --cache_file mnist.cache  > out 2> err
Finished running test 1a
diff -b out /home/jl/programs/vowpal_wabbit/big_tests/expected/1a/out
diff -b err /home/jl/programs/vowpal_wabbit/big_tests/expected/1a/err
8c8
< Reading datafile = /home/jl/programs/vowpal_wabbit/big_tests/dataSets/mnist.dir/train.prep
---
> Reading datafile = /home/melamed/projects/vwDMFork/big_tests/dataSets/mnist.dir/train.prep
12,32c12,32
< 1.000000 1.000000            1            1.0        3        1      125
< 1.000000 1.000000            2            2.0        7        3      192
< 1.000000 1.000000            4            4.0       10        2      128
< 1.000000 1.000000            8            8.0        3       10      206
< 0.937500 0.875000           16           16.0        1        8      197
< 0.718750 0.500000           32           32.0        8        8      140
< 0.671875 0.625000           64           64.0        7        2       91
< 0.562500 0.453125          128          128.0        8        8      100
< 0.445312 0.328125          256          256.0        3        3      174
< 0.322266 0.199219          512          512.0        7        3      144
< 0.265625 0.208984         1024         1024.0        7        7      147
< 0.199707 0.133789         2048         2048.0        1        1      167
< 0.171875 0.144043         4096         4096.0        8        3       81
< 0.147705 0.123535         8192         8192.0        2        2      127
< 0.124817 0.101929        16384        16384.0        8        8      143
< 0.105072 0.085327        32768        32768.0        7        7      166
< 0.086304 0.067535        65536        65536.0       10       10      131
< 0.066536 0.046768       131072       131072.0        6        6      179
< 0.047882 0.029228       262144       262144.0       10       10      113
< 0.031679 0.015476       524288       524288.0        8        8      130
< 0.018461 0.005243      1048576      1048576.0        5        5      104
---
> 1.000000 1.000000            1            1.0        5        1      136
> 1.000000 1.000000            2            2.0        8        5      147
> 0.750000 0.500000            4            4.0        2        5       94
> 0.875000 1.000000            8            8.0        8        5       98
> 0.812500 0.750000           16           16.0        1        1      166
> 0.656250 0.500000           32           32.0       10        2      139
> 0.562500 0.468750           64           64.0        7        7      194
> 0.476562 0.390625          128          128.0        3        3      179
> 0.417969 0.359375          256          256.0        2        2      118
> 0.320312 0.222656          512          512.0        4        4      182
> 0.262695 0.205078         1024         1024.0        2        2      114
> 0.221680 0.180664         2048         2048.0       10       10      132
> 0.177734 0.133789         4096         4096.0        7        7      151
> 0.147217 0.116699         8192         8192.0        2        2       81
> 0.125732 0.104248        16384        16384.0        9        9      176
> 0.104401 0.083069        32768        32768.0        5        5      138
> 0.083542 0.062683        65536        65536.0        2        2       76
> 0.062805 0.042068       131072       131072.0        8        8      151
> 0.044857 0.026909       262144       262144.0        4        4      176
> 0.029736 0.014614       524288       524288.0        1        1      193
> 0.017325 0.004915      1048576      1048576.0        1        1      203
39c39
< average loss = 0.013790
---
> average loss = 0.012981
Test 1a failed due to diffs.
make: *** [1a.compare] Error 1

**Comment by danmelamed** [comment]
The core functionality now works on Ubuntu, Fedora, Cygwin, and Jenkins.  There's more work to do, but the module is already useful.

**Comment by JohnLangford** [comment]
It's still barfing for me (see below). Perhaps you can make this easier 
by _not_ doing a diff?  The point of bigtests is prediction performance 
and time performance, not that every individual printout is precisely 
the same.

-John

  8:40AM humpty-69: make       ~/programs/vowpal_wabbit/big_tests 
[jl/ttypts/24]
Starting to run test 1a
../vowpalwabbit/vw --oaa 10 -d dataSets/mnist.dir/train.prep -f 
runs/1a.dir/mnist.model -b 24 --adaptive --invariant --holdout_off -l 
0.1 --nn 40 --passes 24 -k --compressed --cache_file 
runs/1a.dir/mnist.cache  > runs/1a.dir/out 2> runs/1a.dir/err
Finished running test 1a
diff -b runs/1a.dir/out expected/1a/out
diff -b runs/1a.dir/err expected/1a/err
12,32c12,32
< 1.000000 1.000000            1            1.0        3 1      125
< 1.000000 1.000000            2            2.0        7 3      192
< 1.000000 1.000000            4            4.0       10 2      128
< 1.000000 1.000000            8            8.0        3 10      206
< 0.937500 0.875000           16           16.0        1 8      197
< 0.718750 0.500000           32           32.0        8 8      140
< 0.671875 0.625000           64           64.0        7 2       91
< 0.562500 0.453125          128          128.0        8 8      100
< 0.445312 0.328125          256          256.0        3 3      174
< 0.322266 0.199219          512          512.0        7 3      144
< 0.265625 0.208984         1024         1024.0        7 7      147
< 0.199707 0.133789         2048         2048.0        1 1      167
< 0.171875 0.144043         4096         4096.0        8 3       81
< 0.147705 0.123535         8192         8192.0        2 2      127
< 0.124817 0.101929        16384        16384.0        8 8      143
< 0.105072 0.085327        32768        32768.0        7 7      166
< 0.086304 0.067535        65536        65536.0       10 10      131
< 0.066536 0.046768       131072       131072.0        6 6      179
< 0.047882 0.029228       262144       262144.0       10 10      113
< 0.031679 0.015476       524288       524288.0        8 8      130
< 0.018461 0.005243      1048576      1048576.0        5 5      104
---
 > 1.000000 1.000000            1            1.0       10 1      188
 > 1.000000 1.000000            2            2.0        3 10      137
 > 1.000000 1.000000            4            4.0        4 10      138
 > 1.000000 1.000000            8            8.0        2 3       92
 > 1.000000 1.000000           16           16.0        1 10      154
 > 0.812500 0.625000           32           32.0        7 6      177
 > 0.765625 0.718750           64           64.0        7 7      172
 > 0.617188 0.468750          128          128.0        4 4      147
 > 0.472656 0.328125          256          256.0        2 2       83
 > 0.357422 0.242188          512          512.0        1 1      205
 > 0.274414 0.191406         1024         1024.0        8 5       92
 > 0.222168 0.169922         2048         2048.0        8 8      108
 > 0.177246 0.132324         4096         4096.0        1 1      207
 > 0.150513 0.123779         8192         8192.0        3 3      160
 > 0.127502 0.104492        16384        16384.0        2 2       78
 > 0.106049 0.084595        32768        32768.0        5 5      164
 > 0.085556 0.065063        65536        65536.0        4 4      200
 > 0.064186 0.042816       131072       131072.0        8 8      166
 > 0.045483 0.026779       262144       262144.0        7 7      163
 > 0.030336 0.015190       524288       524288.0        9 9      131
 > 0.017911 0.005486      1048576      1048576.0        4 4      234
39c39
< average loss = 0.013790
---
 > average loss = 0.013497
Test 1a failed due to diffs.
make: *** [1a.compare] Error 1

**Comment by danmelamed** [comment]
Whether we use `diff` or something else, I think we would want to catch discrepancies in the average loss like the one above.  Don't you?

I have tried to eliminate all sources of randomness by comparing results across 4 different platforms, but it looks like there is still some randomness hiding somewhere.  I suggest we track it down together when we can talk in real time.  In the meantime, I will work on adding speed tests, etc.

**Comment by arielf** [comment]
FWIW: this particular diff looks like more than just an accuracy issue.

The actual labels and number of features in the beginning are totally different between reference file and actual.

e.g. In the 1st line of the diff:
```
  < 1.000000 1.000000            1            1.0        3 1      125
  > 1.000000 1.000000            1            1.0       10 1      188
```

**Comment by JohnLangford** [comment]
For the big tests, I think it's a _bug_ if our final prediction performance declines substantially (like .1% relative) or the time on the same platform significantly increases (like 5% relative).  When the text printed out otherwise changes, I don't think we should regard this as a bug.

**Comment by JohnLangford** [comment]
The test suite completed for me :-)  

When it completes, can you print out a summary of performance on each dataset? (and overall performance)?

**Comment by JohnLangford** [comment]
What's the status of this?

**Comment by danmelamed** [comment]
I advertised it on the newsgroup a week ago, but haven’t heard any opinions so far.

**Comment by JohnLangford** [comment]
Status on this?

**Comment by danmelamed** [comment]
The issues that we discussed last time have all been addressed.  I just need to do a bit more testing, and then I will check in the rest of the changes.  Let me know if you want me to prioritize this ahead of the external release project.

**Comment by JohnLangford** [comment]
When I do 'make bigtests' on a fresh clone from the root directory I get the below which seems bad.  Any idea what's up?

<... vw compiling ...> 
make[1]: Entering directory `/media/jl/Last Spinner/temp/vowpal_wabbit/big_tests'
No rules found for 1a.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 1b.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 2.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 3.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 4a.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 4b.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 5a.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 5b.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 6a.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
No rules found for 6b.valid in  Makefile dataSets.mk testDetails.mk testGroups.mk
test suite finished
make[1]: Leaving directory `/media/jl/Last Spinner/temp/vowpal_wabbit/big_tests'

**Comment by danmelamed** [comment]
I just tried it from a fresh checkout on a new machine and saw no problems.  It also works on Jenkins, where every run is from a fresh checkout, albeit using more fine-grained steps.

Can you try it on a machine that I have access to?   That would be a useful data point and also easier to debug.

**Comment by JohnLangford** [comment]
All the Bigtests patch has been merged in.  This is (essentially) a more 
systematic form of the demo directory containing larger scale test which 
are triggered by "make bigtests".  Nothing at presently takes 
particularly long to complete, but it's much greater than what could be 
done with Travis.

-John
